util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,412 @@
+				TimestampList: []uint64{timestamp},
+			}
+			tsr.topSQLMap[string(encodedKey)] = entry
+		} else {
+			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
+			entry.TimestampList = append(entry.TimestampList, timestamp)
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(tsr.topSQLMap) <= tsr.MaxStatementsNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
+	digestCPUTimeList := make([]cpuTimeSort, len(tsr.topSQLMap))
+	{
+		i := 0
+		for key, value := range tsr.topSQLMap {
+			data := cpuTimeSort{
+				Key:            key,
+				SQLDigest:      value.SQLDigest,
+				PlanDigest:     value.PlanDigest,
+				CPUTimeMsTotal: value.CPUTimeMsTotal,
+			}
+			digestCPUTimeList[i] = data
+			i++
+		}
+	}
+
+	// QuickSelect will only return error when the second parameter is out of range
+	if err := quickselect.QuickSelect(cpuTimeSortSlice(digestCPUTimeList), tsr.MaxStatementsNum); err != nil {
+		//	skip eviction
+		return
+
+	}
+
+	// TODO: we can change to periodical eviction (every minute) to relax the CPU pressure
+	shouldEvictList := digestCPUTimeList[tsr.MaxStatementsNum:]
+	for _, evict := range shouldEvictList {
+		delete(tsr.topSQLMap, evict.Key)
+		tsr.mu.Lock()
+		delete(tsr.normalizedSQLMap, string(evict.SQLDigest))
+		delete(tsr.normalizedPlanMap, string(evict.PlanDigest))
+		tsr.mu.Unlock()
+	}
+}
+
+// RegisterSQL registers a normalized SQL string to a SQL digest.
+//
+// Note that the normalized SQL string can be of >1M long.
+// This function should be thread-safe, which means parallelly calling it in several goroutines should be fine.
+// It should also return immediately, and do any CPU-intensive job asynchronously.
+// TODO: benchmark test concurrent performance
+func (tsr *RemoteTopSQLReporter) RegisterSQL(sqlDigest string, normalizedSQL string) {
+	tsr.mu.RLock()
+	_, exist := tsr.normalizedSQLMap[sqlDigest]
+	tsr.mu.RUnlock()
+	if !exist {
+		tsr.mu.Lock()
+		tsr.normalizedSQLMap[sqlDigest] = normalizedSQL
+		tsr.mu.Unlock()
+	}
+}
+
+// RegisterPlan is like RegisterSQL, but for normalized plan strings.
+// TODO: benchmark test concurrent performance
+func (tsr *RemoteTopSQLReporter) RegisterPlan(planDigest string, normalizedPlan string) {
+	tsr.planRegisterChan <- &planRegisterJob{
+		planDigest:     planDigest,
+		normalizedPlan: normalizedPlan,
+	}
+}
+
+// this should be the only place where the normalizedPlanMap is set
+func (tsr *RemoteTopSQLReporter) registerPlanWorker() {
+	// NOTE: if use multiple worker goroutine, we should make sure that access to the digest map is thread-safe
+	for {
+		job := <-tsr.planRegisterChan
+		tsr.mu.RLock()
+		_, exist := tsr.normalizedPlanMap[job.planDigest]
+		tsr.mu.RUnlock()
+		if exist {
+			continue
+		}
+		planDecoded, err := tsr.planBinaryDecoder(job.normalizedPlan)
+		if err != nil {
+			fmt.Printf("decode plan failed: %v\n", err)
+			continue
+		}
+		tsr.normalizedPlanMap[job.planDigest] = planDecoded
```

Comment:
This need to be protected behind a lock.

Reply:
fixed